lib/puppet/type/zfs_share.rb

```diff
@@ -0,0 +1,234 @@
+Puppet::Type.newtype(:zfs_share) do
+  @doc = "Manage zfs shares"
+
+  require 'puppet/parameter/boolean'
+
+  ensurable
+
+  # zfs share documentation
+  # https://docs.oracle.com/cd/E26502_01/html/E29031/zfs-share-1m.html
+
+  newparam(:name) do
```

Comment:
I like the idea, but I have some trouble when working through it. Name is consistently used to refer to the volume in question, and to refer to it as something else requires the other components be referred to differently. It seems more logical to keep the named share component separate when considering named shares a option of a share, which is the name. The named share ends up being collected as a component of the share when looking at all of the shares.